File: app/src/main/java/com/example/audio/SpookyAudioEngine.kt
package com.example.audio

import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import kotlinx.coroutines.*
import kotlin.math.sin

class SpookyAudioEngine {
    private var bgmTrack: AudioTrack? = null
    private var sfxTrack: AudioTrack? = null
    private var bgmJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    private var isBgmActive = false
    private var bgmType: String? = null

    fun startBgm(type: String) {
        stopBgm()
        isBgmActive = true
        bgmType = type
        bgmJob = scope.launch {
            val sampleRate = 22050
            val bufferSize = AudioTrack.getMinBufferSize(
                sampleRate,
                AudioFormat.CHANNEL_OUT_MONO,
                AudioFormat.ENCODING_PCM_16BIT
            )
            val track = AudioTrack(
                AudioManager.STREAM_MUSIC,
                sampleRate,
                AudioFormat.CHANNEL_OUT_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
                bufferSize,
                AudioTrack.MODE_STREAM
            )
            bgmTrack = track
            try {
                track.play()
            } catch (e: Exception) {
                return@launch
            }

            val buffer = ShortArray(bufferSize)
            var time = 0.0

            while (isActive && isBgmActive) {
                for (i in buffer.indices) {
                    time += 1.0 / sampleRate
                    // Low multi-tone spooky drone: combined low frequency sine waves
                    val freq1 = 55.0
                    val freq2 = 57.0
                    val modulation = 0.55 + 0.45 * sin(2.0 * Math.PI * 0.15 * time) // Swell effect
                    
                    val value = (sin(2.0 * Math.PI * freq1 * time) + sin(2.0 * Math.PI * freq2 * time)) / 2.0
                    val floatSample = value * 0.25 * modulation
                    
                    // Add subtle crackle/hiss for realistic atmosphere
                    val noise = (Math.random() * 2.0 - 1.0) * 0.012
                    
                    val finalSample = (floatSample + noise).coerceIn(-1.0, 1.0)
                    buffer[i] = (finalSample * Short.MAX_VALUE).toInt().toShort()
                }
                try {
                    track.write(buffer, 0, buffer.size)
                } catch (e: Exception) {
                    break
                }
            }
            try {
                track.stop()
                track.release()
            } catch (e: Exception) {}
        }
    }

    fun stopBgm() {
        isBgmActive = false
        bgmJob?.cancel()
        bgmJob = null
        bgmTrack?.let {
            try {
                it.stop()
                it.release()
            } catch (e: Exception) {}
        }
        bgmTrack = null
        bgmType = null
    }

    fun playSfx(type: String) {
        scope.launch {
            val sampleRate = 22050
            val durationSeconds = when {
                type.contains("Scream", ignoreCase = true) -> 1.5
                type.contains("Creak", ignoreCase = true) -> 1.3
                type.contains("Wind", ignoreCase = true) || type.contains("Rain", ignoreCase = true) -> 2.0
                type.contains("Heartbeat", ignoreCase = true) -> 1.6
                type.contains("Whisper", ignoreCase = true) -> 1.8
                type.contains("Shatter", ignoreCase = true) -> 0.8
                else -> 1.0
            }
            val numSamples = (sampleRate * durationSeconds).toInt()
            val buffer = ShortArray(numSamples)

            when {
                type.contains("Scream", ignoreCase = true) -> {
                    // Eerie rising and collapsing vibration scream
                    for (i in 0 until numSamples) {
                        val t = i.toDouble() / sampleRate
                        val freq = 700.0 + 250.0 * sin(2.0 * Math.PI * 30.0 * t) - 300.0 * t
                        val fadeOut = (1.0 - t / durationSeconds).coerceIn(0.0, 1.0)
                        val noise = (Math.random() * 2.0 - 1.0) * 0.12
                        val sweep = sin(2.0 * Math.PI * freq * t) * 0.20
                        val sample = (sweep + noise) * fadeOut
                        buffer[i] = (sample * Short.MAX_VALUE).toInt().toShort()
                    }
                }
                type.contains("Creak", ignoreCase = true) -> {
                    // Unstable, sliding, friction frequency
                    for (i in 0 until numSamples) {
                        val t = i.toDouble() / sampleRate
                        val baseFreq = 160.0
                        val jitter = sin(2.0 * Math.PI * 55.0 * t) * 25.0
                        val freq = baseFreq + jitter + (t * 40.0)
                        val amplitude = (t * (durationSeconds - t) * 1.5).coerceIn(0.0, 1.0)
                        val sample = sin(2.0 * Math.PI * freq * t) * 0.14 * amplitude
                        buffer[i] = (sample * Short.MAX_VALUE).toInt().toShort()
                    }
                }
                type.contains("Heartbeat", ignoreCase = true) -> {
                    // Double heavy pulse of low frequency
                    for (i in 0 until numSamples) {
                        val t = i.toDouble() / sampleRate
                        val pulse1 = sin(2.0 * Math.PI * 50.0 * t) * Math.exp(-16.0 * t)
                        val t2 = t - 0.45
                        val pulse2 = if (t2 > 0) sin(2.0 * Math.PI * 50.0 * t2) * Math.exp(-16.0 * t2) else 0.0
                        val sample = (pulse1 + pulse2 * 0.85) * 0.55
                        buffer[i] = (sample * Short.MAX_VALUE).toInt().toShort()
                    }
                }
                type.contains("Whisper", ignoreCase = true) -> {
                    // Distorted whispering wind
                    for (i in 0 until numSamples) {
                        val t = i.toDouble() / sampleRate
                        val noise = (Math.random() * 2.0 - 1.0) * 0.08
                        val hiss = sin(2.0 * Math.PI * (1200.0 + 300.0 * sin(2.0 * Math.PI * 4.0 * t)) * t) * 0.015
                        val fadeOut = (1.0 - t / durationSeconds).coerceIn(0.0, 1.0)
                        val sample = (noise + hiss) * fadeOut
                        buffer[i] = (sample * Short.MAX_VALUE).toInt().toShort()
                    }
                }
                type.contains("Shatter", ignoreCase = true) -> {
                    // Rapid decay metallic noise
                    for (i in 0 until numSamples) {
                        val t = i.toDouble() / sampleRate
                        val noise = (Math.random() * 2.0 - 1.0) * 0.15
                        val ring = sin(2.0 * Math.PI * 1800.0 * t) * 0.05
                        val sample = (noise + ring) * Math.exp(-22.0 * t)
                        buffer[i] = (sample * Short.MAX_VALUE).toInt().toShort()
                    }
                }
                else -> { // Default howling wind / rain rumble
                    for (i in 0 until numSamples) {
                        val t = i.toDouble() / sampleRate
                        val noise = (Math.random() * 2.0 - 1.0) * 0.05
                        val rum = sin(2.0 * Math.PI * 70.0 * t) * 0.03
                        val sample = (noise + rum) * (1.0 - t / durationSeconds)
                        buffer[i] = (sample * Short.MAX_VALUE).toInt().toShort()
                    }
                }
            }

            withContext(Dispatchers.IO) {
                try {
                    val track = AudioTrack(
                        AudioManager.STREAM_MUSIC,
                        sampleRate,
                        AudioFormat.CHANNEL_OUT_MONO,
                        AudioFormat.ENCODING_PCM_16BIT,
                        numSamples * 2,
                        AudioTrack.MODE_STATIC
                    )
                    track.write(buffer, 0, buffer.size)
                    track.play()
                    delay((durationSeconds * 1000).toLong())
                    track.stop()
                    track.release()
                } catch (e: Exception) {}
            }
        }
    }

    fun shutdown() {
        stopBgm()
        scope.cancel()
    }
}

File: lib/audio/spooky_audio_engine.dart
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:audioplayers/audioplayers.dart';

/// Procedurally synthesizes the background drone and one-shot sound effects as
/// PCM/WAV and plays them through audioplayers — the `SpookyAudioEngine`
/// equivalent (Android's raw `AudioTrack` has no Flutter counterpart, so we
/// build WAV byte buffers instead).
class SpookyAudioEngine {
  static const int _sampleRate = 22050;
  final math.Random _rng = math.Random();

  final AudioPlayer _bgmPlayer = AudioPlayer(playerId: 'spooky_bgm');
  final AudioPlayer _sfxPlayer = AudioPlayer(playerId: 'spooky_sfx');

  bool _bgmActive = false;

  // --- Background music: low multi-tone spooky drone, looped ---

  Future<void> startBgm(String type) async {
    await stopBgm();
    _bgmActive = true;

    // One full swell-modulation period (0.15 Hz -> ~6.667s) loops seamlessly.
    const double durationSeconds = 1.0 / 0.15;
    final int numSamples = (_sampleRate * durationSeconds).round();
    final samples = Int16List(numSamples);

    for (var i = 0; i < numSamples; i++) {
      final t = i / _sampleRate;
      const freq1 = 55.0;
      const freq2 = 57.0;
      final modulation = 0.55 + 0.45 * math.sin(2.0 * math.pi * 0.15 * t);
      final value =
          (math.sin(2.0 * math.pi * freq1 * t) +
              math.sin(2.0 * math.pi * freq2 * t)) /
          2.0;
      final floatSample = value * 0.25 * modulation;
      final noise = (_rng.nextDouble() * 2.0 - 1.0) * 0.012;
      samples[i] = _toPcm16(floatSample + noise);
    }

    try {
      await _bgmPlayer.setReleaseMode(ReleaseMode.loop);
      await _bgmPlayer.play(BytesSource(_buildWav(samples), mimeType: 'audio/wav'));
    } catch (_) {
      // Ignore playback failures (e.g. no audio device).
    }
  }

  Future<void> stopBgm() async {
    _bgmActive = false;
    try {
      await _bgmPlayer.stop();
    } catch (_) {}
  }

  bool get isBgmActive => _bgmActive;

  // --- One-shot sound effects ---

  Future<void> playSfx(String type) async {
    final lower = type.toLowerCase();
    final double durationSeconds;
    if (lower.contains('scream')) {
      durationSeconds = 1.5;
    } else if (lower.contains('creak')) {
      durationSeconds = 1.3;
    } else if (lower.contains('wind') || lower.contains('rain')) {
      durationSeconds = 2.0;
    } else if (lower.contains('heartbeat')) {
      durationSeconds = 1.6;
    } else if (lower.contains('whisper')) {
      durationSeconds = 1.8;
    } else if (lower.contains('shatter')) {
      durationSeconds = 0.8;
    } else {
      durationSeconds = 1.0;
    }

    final numSamples = (_sampleRate * durationSeconds).round();
    final samples = Int16List(numSamples);

    for (var i = 0; i < numSamples; i++) {
      final t = i / _sampleRate;
      double sample;
      if (lower.contains('scream')) {
        final freq = 700.0 + 250.0 * math.sin(2.0 * math.pi * 30.0 * t) - 300.0 * t;
        final fadeOut = (1.0 - t / durationSeconds).clamp(0.0, 1.0);
        final noise = (_rng.nextDouble() * 2.0 - 1.0) * 0.12;
        final sweep = math.sin(2.0 * math.pi * freq * t) * 0.20;
        sample = (sweep + noise) * fadeOut;
      } else if (lower.contains('creak')) {
        const baseFreq = 160.0;
        final jitter = math.sin(2.0 * math.pi * 55.0 * t) * 25.0;
        final freq = baseFreq + jitter + (t * 40.0);
        final amplitude = (t * (durationSeconds - t) * 1.5).clamp(0.0, 1.0);
        sample = math.sin(2.0 * math.pi * freq * t) * 0.14 * amplitude;
      } else if (lower.contains('heartbeat')) {
        final pulse1 = math.sin(2.0 * math.pi * 50.0 * t) * math.exp(-16.0 * t);
        final t2 = t - 0.45;
        final pulse2 =
            t2 > 0 ? math.sin(2.0 * math.pi * 50.0 * t2) * math.exp(-16.0 * t2) : 0.0;
        sample = (pulse1 + pulse2 * 0.85) * 0.55;
      } else if (lower.contains('whisper')) {
        final noise = (_rng.nextDouble() * 2.0 - 1.0) * 0.08;
        final hiss = math.sin(
              2.0 * math.pi * (1200.0 + 300.0 * math.sin(2.0 * math.pi * 4.0 * t)) * t,
            ) *
            0.015;
        final fadeOut = (1.0 - t / durationSeconds).clamp(0.0, 1.0);
        sample = (noise + hiss) * fadeOut;
      } else if (lower.contains('shatter')) {
        final noise = (_rng.nextDouble() * 2.0 - 1.0) * 0.15;
        final ring = math.sin(2.0 * math.pi * 1800.0 * t) * 0.05;
        sample = (noise + ring) * math.exp(-22.0 * t);
      } else {
        // Default howling wind / rain rumble.
        final noise = (_rng.nextDouble() * 2.0 - 1.0) * 0.05;
        final rum = math.sin(2.0 * math.pi * 70.0 * t) * 0.03;
        sample = (noise + rum) * (1.0 - t / durationSeconds);
      }
      samples[i] = _toPcm16(sample);
    }

    try {
      await _sfxPlayer.stop();
      await _sfxPlayer.setReleaseMode(ReleaseMode.stop);
      await _sfxPlayer.play(BytesSource(_buildWav(samples), mimeType: 'audio/wav'));
    } catch (_) {}
  }

  Future<void> shutdown() async {
    await stopBgm();
    await _bgmPlayer.dispose();
    await _sfxPlayer.dispose();
  }

  // --- Helpers ---

  int _toPcm16(double sample) {
    final clamped = sample.clamp(-1.0, 1.0);
    return (clamped * 32767).round();
  }

  /// Wraps mono 16-bit PCM samples in a minimal RIFF/WAV container.
  Uint8List _buildWav(Int16List samples) {
    const channels = 1;
    const bitsPerSample = 16;
    final byteRate = _sampleRate * channels * bitsPerSample ~/ 8;
    final blockAlign = channels * bitsPerSample ~/ 8;
    final dataSize = samples.length * 2;
    final fileSize = 44 + dataSize;

    final bytes = ByteData(fileSize);
    var offset = 0;

    void writeString(String s) {
      for (final unit in s.codeUnits) {
        bytes.setUint8(offset++, unit);
      }
    }

    void writeUint32(int v) {
      bytes.setUint32(offset, v, Endian.little);
      offset += 4;
    }

    void writeUint16(int v) {
      bytes.setUint16(offset, v, Endian.little);
      offset += 2;
    }

    writeString('RIFF');
    writeUint32(fileSize - 8);
    writeString('WAVE');
    writeString('fmt ');
    writeUint32(16); // PCM fmt chunk size
    writeUint16(1); // audio format = PCM
    writeUint16(channels);
    writeUint32(_sampleRate);
    writeUint32(byteRate);
    writeUint16(blockAlign);
    writeUint16(bitsPerSample);
    writeString('data');
    writeUint32(dataSize);

    for (final s in samples) {
      bytes.setInt16(offset, s, Endian.little);
      offset += 2;
    }

    return bytes.buffer.asUint8List();
  }
}
